statistics/histogram.go

```diff
@@ -826,12 +826,108 @@ func (hg *Histogram) AvgCountPerNotNullValue(totalCount int64) float64 {
 
 func (hg *Histogram) outOfRange(val types.Datum) bool {
 	if hg.Len() == 0 {
-		return true
+		return false
 	}
 	return chunk.Compare(hg.Bounds.GetRow(0), 0, &val) > 0 ||
 		chunk.Compare(hg.Bounds.GetRow(hg.Bounds.NumRows()-1), 0, &val) < 0
 }
 
+// outOfRangeRowCount estimate the row count of part of [lDatum, rDatum] which is out of range of the histogram.
+// Here we assume the density of data is decreasing from the lower/upper bound of the histogram toward outside.
+// The maximum row count it can get is the increaseCount. It reaches the maximum when out-of-range width reaches histogram range width.
+// As it shows below. To calculate the out-of-range row count, we need to calculate the percentage of the shaded area.
+//
+//               /│             │\
+//             /  │             │  \
+//           /x│  │◄─histogram─►│    \
+//         / xx│  │    range    │      \
+//       / │xxx│  │             │        \
+//     /   │xxx│  │             │          \
+//────┴────┴───┴──┴─────────────┴───────────┴─────
+//    ▲    ▲   ▲  ▲             ▲           ▲
+//    │    │   │  │             │           │
+// boundL  │   │histL         histR       boundR
+//         │   │
+//    lDatum  rDatum
+func (hg *Histogram) outOfRangeRowCount(lDatum, rDatum *types.Datum, increaseCount int64) float64 {
+	if hg.Len() == 0 {
+		return 0
+	}
+
+	// For bytes and string type, we need to cut the common prefix when converting them to scalar value.
+	// Here we calculate the length of common prefix.
+	commonPrefix := 0
+	if hg.GetLower(0).Kind() == types.KindBytes || hg.GetLower(0).Kind() == types.KindString {
+		// Calculate the common prefix length among the lower and upper bound of histogram and the range we want to estimate.
+		commonPrefix = commonPrefixLength(hg.GetLower(0).GetBytes(), hg.GetUpper(hg.Len()-1).GetBytes())
+		commonPrefix2 := commonPrefixLength(lDatum.GetBytes(), rDatum.GetBytes())
+		if commonPrefix2 < commonPrefix {
+			commonPrefix = commonPrefix2
+		}
+		commonPrefix3 := commonPrefixLength(hg.GetLower(0).GetBytes(), lDatum.GetBytes())
+		if commonPrefix3 < commonPrefix {
+			commonPrefix = commonPrefix3
+		}
+	}
+
+	// Convert the range we want to estimate to scalar value(float64)
+	l := convertDatumToScalar(lDatum, commonPrefix)
+	r := convertDatumToScalar(rDatum, commonPrefix)
+	// make sure l < r
+	if l >= r {
+		return 0
+	}
+	// Convert the lower and upper bound of the histogram to scalar value(float64)
+	histL := convertDatumToScalar(hg.GetLower(0), commonPrefix)
+	histR := convertDatumToScalar(hg.GetUpper(hg.Len()-1), commonPrefix)
+	histWidth := histR - histL
+	if histWidth <= 0 {
+		return 0
+	}
+	boundL := histL - histWidth
+	boundR := histR + histWidth
+
+	leftPercent := float64(0)
+	rightPercent := float64(0)
+
+	// keep l and r unchanged, use actualL and actualR to calculate.
+	actualL := l
+	actualR := r
+	// Handling the out-of-range part on the left of the histogram range
+	if actualL < histL && actualR > boundL {
```

Comment:
How about `actualL >= histL || actualR <= boundL`?

Reply:
I didn't get it.
I think they are reversed.